### PR TITLE
search: add top queries to usage stats page

### PR DESCRIPTION
**Pull request by ijsnow**
Adds the top queries to the usage stats page for site admins. Note this PR is based on https://github.com/sourcegraph/sourcegraph/pull/3818.

Screenshot 2019-05-06 10 24 31

### CODE REVIEW

**Comment by attfarhan** [comment]
@ijt I fixed the build for this. We can merge this now if you still want to get this in.

**Comment by ijt** [comment]
I'd like to, except there seems to be a problem with the CSS. When I tried this locally, the input box for the number of queries to show was hiding the number:

Screen Shot 2019-05-22 at 21 59 03

I tweaked the CSS in the developer tools and got the number to appear, like this:

Screen Shot 2019-05-22 at 22 01 46

Can you fix the UI before merging?

**Comment by dadlerj** [comment]
Does this data live over all time? Or is it "in the past month" or something? That should be clear in the UI!

Also, the placement of this is a bit weird to me. There will three sections after this is added: 
1) a chart with counts of all active users
2) a list of details of all active users, tied to the same timeframes as the charts above
3) a list of top searches

Why would 3) be put in between 1) and 2) (which are logically connected as related to end users)? In my mind, this should be an entirely different page, since I imagine ultimately it will provide more slicing and dicing options (e.g. show the top 10 repos searched over, the top 10 searches in repo X only, etc.).

**Comment by ijt** [comment]
> Does this data live over all time? Or is it "in the past month" or something? That should be clear in the UI!

It's the most recent search results, up to 1e5 of them. That's currently hard-coded, but we should have something to configure it.

File: web/src/site-admin/SearchUsageStats.scss [hunk 1]
```diff
@@ -0,0 +1,32 @@
+.search-usage-stats {
+    &__header {
+        display: flex;
+
+        &--title {
```
Comment (felixfbecker) [inline web/src/site-admin/SearchUsageStats.scss]: `--` are modifiers. This is not a modifier, it's another element (`.search-usage-stats__title`).

File: web/src/site-admin/SearchUsageStats.scss [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+            flex: 1;
+        }
+        &--input {
```
Comment (felixfbecker) [inline web/src/site-admin/SearchUsageStats.scss]: Same

File: web/src/site-admin/SearchUsageStats.scss [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            flex: 0;
+        }
+    }
+
+    &__top-queries {
+        width: 100%;
+
+        &--entry {
```
Comment (felixfbecker) [inline web/src/site-admin/SearchUsageStats.scss]: Same

File: web/src/site-admin/SearchUsageStats.scss [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+            .theme-light & {
+                border-bottom: 1px solid $table-border-color-light;
+            }
+
+            &:last-of-type {
+                border-bottom: none;
+            }
+
+            &-cell {
```
Comment (felixfbecker) [inline web/src/site-admin/SearchUsageStats.scss]: Elements shouldn't be nested further, should just be `__cell` and `__row`.

**Comment by christinaforney** [comment]
@ijt - who is the proper owner of this change? It sounds like (based on discussion above) it is considered useful, though there is code feedback and UI changes requested that need to be handled. Not necessarily saying to prioritize this, but want to make sure it doesn't get too stale to be relevant. Thoughts?

**Comment by christinaforney** [comment]
@ijt - who is the proper owner of this change? It sounds like (based on discussion above) it is considered useful, though there is code feedback and UI changes requested that need to be handled. Not necessarily saying to prioritize this, but want to make sure it doesn't get too stale to be relevant. Thoughts?

Actions:
- [ ] Address feedback from @felixfbecker 
- [ ] Fix the input box
- [ ] Move top search results below registered users (does registered users always limit to top 20?)

**Comment by dadlerj** [comment]
> Move top search results below registered users (does registered users always limit to top 20?)

No, you can expand to the full user list. We have historically always had one "filterable" list per page (see users, repos, orgs, survey responses, usage analytics, etc.), and I don't think the UI would work very well with two... I wouldn't stand in the way of this, I just want to call out that it'd be an odd UI (and I'm not even sure how the URL params like `?first=` would work here — would they collide?).

**Comment by ijt** [comment]
@christinaforney, we still need to add something to prevent automated queries from getting into the `recent_searches` table before it would make sense to expose that table's contents within the UI.

@attfarhan would probably be the one to own the frontend changes.

**Comment by christinaforney** [comment]
Thanks @ijt! I just pushed changes to address comments and fix the input box. Depending on other necessary changes (if more involved) @attfarhan can own.

👍 to needing to fix the `recent_searches` table first.

Another question I have: Why do we allow for changing the number of displayed items in the table in the first place? It doesn't seem like these features considered scale (how does it work with thousands of queries or registered users). I think it would be worthwhile to spend some time thinking about the general design of this page, but this could go in as is.

**Comment by dadlerj** [comment]
>Another question I have: Why do we allow for changing the number of displayed items in the table in the first place? It doesn't seem like these features considered scale (how does it work with thousands of queries or registered users). I think it would be worthwhile to spend some time thinking about the general design of this page, but this could go in as is.

Great question. Typically these lists are filterable — e.g. on the repo page you can filter to only "cloned", etc, or you can search for a name. Same on the users page. This all depends on the backend's ability to execute this sort of filtering though... E.g. for repos, we can just construct a SQL query that adds a `WHERE name like %${search_text}%`. For usage stats, it's a bit more technically challenging (since they are a combination of user accounts from SQL and usage stats stored in Redis). Instead, we provide a few built-in filters (such as "active in the last day/week/month").

Not sure if there's a better way to do this in general, though, but if you have any ideas you should share. It's a similar philosophy to our own search product — you can always see more if you need to, but making it quickly filterable is ideal.

**Comment by christinaforney** [comment]
Thanks @dadlerj !

> Not sure if there's a better way to do this in general, though, but if you have any ideas you should share. It's a similar philosophy to our own search product — you can always see more if you need to, but making it quickly filterable is ideal.

That makes sense - and I support the filterable concept. The challenge here is: what would an admin want to learn from top search queries? Is there any reason this should be filterable? Should they be able to look for a particular query? Or is it truly just a "show me the most popular ones". What information is this intended to communicate to the admin? What are they supposed to do with this info?

Some ideas to make this more useful: add a count of unique users who have run the "top queries" - when you have those two numbers you can learn more - I have one user doing this a lot vs. I have many users doing this.

**Comment by slimsag** [comment]
This PR has been open for a very long time with no progress, closing. Reopen if desired.

**Pull request closed without merging**